File: python-equivalent/wer.py
import sys

# Credits: https://martin-thoma.com/word-error-rate-calculation/
def wer(h, r):
    """
    Calculation of WER with Levenshtein distance.

    Works only for iterables up to 254 elements (uint8).
    O(nm) time ans space complexity.

    Parameters
    ----------
    r : list
    h : list

    Returns
    -------
    int

    Examples
    --------
    >>> wer("who is there".split(), "is there".split())
    1
    >>> wer("who is there".split(), "".split())
    3
    >>> wer("".split(), "who is there".split())
    3
    """
    # initialisation
    import numpy
    if len(h) == 0:
        return len(r)

    d = numpy.zeros((len(r) + 1) * (len(h) + 1), dtype=numpy.uint8)
    d = d.reshape((len(r) + 1, len(h) + 1))
    for i in range(len(r) + 1):
        for j in range(len(h) + 1):
            if i == 0:
                d[0][j] = j
            elif j == 0:
                d[i][0] = i

    # computation
    for i in range(1, len(r) + 1):
        for j in range(1, len(h) + 1):
            if r[i - 1] == h[j - 1]:
                d[i][j] = d[i - 1][j - 1]
            else:
                substitution = d[i - 1][j - 1] + 1
                insertion = d[i][j - 1] + 1
                deletion = d[i - 1][j] + 1
                d[i][j] = min(substitution, insertion, deletion)

    return d[len(r)][len(h)]


if __name__ == '__main__':
    f1 = sys.argv[1]
    f2 = sys.argv[2]
    with open(f1, 'r', encoding='utf-8') as f:
        preds = f.readlines()
    with open(f2, 'r', encoding='utf-8') as f:
        truth = f.readlines()
    wers = [wer(l1.split(), l2.split())/len(l2.split()) for l1, l2 in zip(preds, truth)]
    print("FINAL WER:", sum(wers)/len(wers))

File: julia-equivalent/wer.jl
#=
wer:
- Julia version: 
- Author: geoph
- Date: 2021-02-06
=#

# Credits: https://martin-thoma.com/word-error-rate-calculation/
function wer(h, r)::Float32
    """
    Calculation of WER with Levenshtein distance.

    Works only for iterables up to 254 elements (uint8).
    O(nm) time ans space complexity.

    Parameters
    ----------
    r : Array of true words
    h : Array of predicted words

    Returns
    -------
    float32
    """
    println("H: ", h)
    println("R: ", r)
    if length(h) == 0
        return 1f0
    end

    d = zeros(UInt8, (length(r) + 1), (length(h) + 1))
    for i = 1:(length(r) + 1)
        for j = 1:(length(h) + 1)
            if i == 1
                d[1, j] = j
            elseif j == 1
                d[i, 1] = i
            end
        end
    end
    # computation
    for i = 2:(length(r) + 1)
        for j = 2:(length(h) + 1)
            if r[i - 1] == h[j - 1]
                d[i, j] = d[i - 1, j - 1]
            else
                substitution = d[i - 1, j - 1] + 1
                insertion = d[i, j - 1] + 1
                deletion = d[i - 1, j] + 1
                d[i, j] = min(substitution, insertion, deletion)
            end
        end
    end
    Float32(d[length(r), length(h)]) / Float32(length(r))
end

predictions = ARGS[1]
truth = ARGS[2]

my_transcriptions = []
original_transcript = []

open(predictions, "r") do f
    for line in readlines(f)
        push!(my_transcriptions, split(line, " "))
    end
end

open(predictions, "r") do f
    for line in readlines(f)
        push!(original_transcript, split(line, " "))
    end
end

if length(my_transcriptions) != length(original_transcript)
    println("Invalid lengths")
    exit(1)
end

println("Calculating wers...")
wers = [wer(i[1], i[2]) for i in zip(my_transcriptions, original_transcript)]
println(wers)
final_wer = sum(wers) / length(wers)
println("Output WER: ", final_wer)
